### PR TITLE
feat: push notifications

### PR DESCRIPTION
**Pull request by sgammon**
## Summary

Adds initial support wiring for Push Notifications (see: #11651).

This requires instrumentation down into `tao`, to call methods on the `AppDelegate` or `NSApplication` / `UIApplication` (for Apple platforms). Firebase Messaging (FCM) support is also under consideration.

- [ ] **`tauri`**: Handle new events
  - (that's this pr)
  - [x] Forward/map `PushRegistration` event
  - [x] Forward/map `PushRegistrationFailed` event
  - [ ] General cleanup (don't repeat typedefs, etc)
- [ ] **`tao`**: Hooks into `UIApplication` / `NSApplication`
  - tauri-apps/tao#1015
  - [x] Support for macOS
  - [ ] Support for iOS
  - [x] Any considerations for Android 
  - [x] Considerations for Windows
- [ ] **`tauri-plugin-push-notifications`**:
  - tauri-apps/plugins-workspace#2066
  - [x] Initially working plugin structure ([here](https://github.com/sgammon/tauri-plugin-push-notifications))
  - [x] Facade for `await pushToken(...)`
  - [x] Working local test to capture token
  - [x] Add plugin sources to `plugins-workspace`
  - [ ] Interface for callback-based token access
- [ ] **`tauri-docs`**
  - tauri-apps/tauri-docs#2997 
  - [x] Add initial Push Notifications plugin page
  - [ ] Validate configuration with example app
- [ ] **Platform support**
  - [x] macOS (APNS)
  - [ ] iOS (APNS)
  - [ ] Windows (WNS)
  - [x] Android (FCM)

### PR Tree

This PR is accompanied by a plugin, `tauri-plugin-push-notifications`, and a PR to `tao`:

- [Plugin repo](https://github.com/sgammon/tauri-plugin-push-notifications)
- tauri-apps/tao#1015
- tauri-apps/plugins-workspace#2066
- tauri-apps/tauri-docs#2997

### Known Issues

- [ ] **General**
  - [ ] Token needs strong typing in guest code
- [ ] **Docs**
  - [ ] macOS and iOS devices need an entitlement, which should be mentioned in the docs.
- [ ] **Tests**
  - [ ] There are no tests yet.

> [!WARNING]
> This functionality is experimental and is gated behind the `push-notifications` feature for now.

### How it works

(1) `tao` and `tauri` forward two new events: `PushRegistration(PushToken)` and `PushRegistrationFailed(Error)`.
(2) a new plugin, `tauri-plugin-push-notifications`, listens for these events, holds the token or error, and exposes those values to guest code when requested.

```mermaid
sequenceDiagram
    tao->>+OS: Setup app delegate
    OS->>+tao: applicationDidFinishLaunching
    tao->>+OS: registerForRemoteNotifications()
    OS->>+tao: didFinishRegisteringWithToken()
    tao->>+tauri: PushRegistration(PushToken) Event
    tauri->>+Plugin: Forward event
    Plugin->>+Plugin: Retain token in-memory
    tauri->>+App: Launch app
    App<<->>+Plugin: await pushToken()
```

### CODE REVIEW

**Comment by ChrisCates** [comment]
Any way I can help ship this?

**Comment by ChrisCates** [comment]
Any way I can help ship this?
That or, what's your ETA for completion brotha?

**Comment by sgammon** [comment]
@ChrisCates Very close to completion, i'm testing on macOS shortly. I haven't explored Android yet, so some research there might help (how best to integrate Firebase Cloud Messaging - FCM - into `tao` and `tauri`; ideally it could work within the plugin only).

Windows is looking solid but I'm not sure how to extract the token yet. That may remain experimental until later if only because I can't find any third party push services that support WNS at this time.

tl;dr I'd say we could get macOS and potentially iOS working within a few days. Android and WNS are probably easier/less complex, but may take a few days longer for research about how best to integrate. Regarding review and merge, I've no clue, this is my first PR to Tauri.

**Comment by sgammon** [comment]
this needs a rebase but otherwise it is close to being done. there also need to be methods to trigger a request for notification permissions, which is required on platforms like iOS and Android.

windows notification service ("WNS") will probably remain experimental for now

**Comment by sgammon** [comment]
cc / @amrbashir Could somebody on the Tauri team rendezvous with me on Discord (I'm in the contributors channel), or perhaps offer some guidance about whether this PR could be welcome?

**Comment by amrbashir** [comment]
I don't think there are any objections around the added functionality, it is just this PR is marked as draft so the team probably thought it is not finished yet.

That said, it looks good from my perspective but since this PR is more specific to macOS/iOS, I'd better let @lucasfernog or @FabianLars review and approve 

Also feel free to ping with us in the Discord contributors channel if you need to dicuss something

**Comment by sgammon** [comment]
Understood @amrbashir and thank you for getting back so quickly. It's marked as a draft because it will involve changes to (at least) `tao` and the plugin workspace; I can try to minimize the impact of the PR in general especially with guidance about iOS.

I guess I'm worried that this is invasive enough that it may be a non-starter. I'm glad to see a positive nod. I can rebase this and get it as clean as possible, avoiding speculative changes, and then mark for review

(Still, any preliminary feedback is welcome)

**Comment by amrbashir** [comment]
Don't worry about how many crates you need to change, it is something we do often.

**Comment by RyanBrewer317** [comment]
So is this done enough that I could use this branch for an android app with notifications? (I'm not patient enough to wait for it to get merged lol)
